lib/openshift/cluster_version.rb

```diff
@@ -15,5 +15,55 @@ def version(user: nil, cached: true, quiet: false)
       rr = raw_resource(user: user, cached: cached, quiet: quiet)
       rr.dig('status', 'desired', 'version')
     end
+
+    def history(user: nil, cached: false, quiet: false)
+      rr = raw_resource(user: user, cached: cached, quiet: quiet)
+      rr.dig('status', 'history')
+    end
+
+    # return a list of history matching the state we want to filter on
+    # query_hash is the parameter we want to filter on,
+    def history_matching(key: nil, value:nil , user: nil, cached: true, quiet: false)
+      self.history.select { |h| h[key] == value }
+    end
+
+    def image(user: nil, cached: true, quiet: false)
+      rr = raw_resource(user: user, cached: cached, quiet: quiet)
+      rr.dig('status', 'desired', 'image')
+    end
+
+    # get the last entry of the spec.history which should be where the original
+    def initial_image_url(user: nil, cached: true, quiet: false)
+      self.history.last['image'].split('@sha256:')[0]
+    end
+
+    # @param version [String] upgrade version that we check for completion
+    def upgrade_completed?(version:, user: nil, cached: false, quiet: false)
+      res = self.history_matching(key: 'version', value: version)
+      if res.count > 0
+        logger.debug("MSG: #{res.first['message']}")
+        res.first['state'] == 'Completed'
+      else
+        false
+      end
+    end
+
+    def wait_for_upgrade_completion(version:, user: nil, cached: false, quiet: false, upgrade_timeout: 2*60*60)
```

Comment:
did we talk about `s/upgrade_timeout/timeout/` or remove that parameter altogether?

Reply:
I changed it to `timeout`